File: solutions/kotlin/robot-simulator/1/src/main/kotlin/Robot.kt
class Robot(var gridPosition: GridPosition = GridPosition(0, 0),
            var orientation: Orientation = Orientation.NORTH) {

    private val orientationCount = Orientation.values().size

    fun simulate(instructions: String) {
        for (command in instructions) {
           when(command) {
               'L' -> orientation = Orientation.values()[(orientation.ordinal - 1 + orientationCount) % orientationCount]
               'R' -> orientation = Orientation.values()[(orientation.ordinal + 1) % orientationCount]
               'A' -> {
                   gridPosition = when(orientation) {
                       Orientation.NORTH -> GridPosition(gridPosition.x, gridPosition.y + 1)
                       Orientation.EAST -> GridPosition(gridPosition.x + 1, gridPosition.y)
                       Orientation.SOUTH -> GridPosition(gridPosition.x, gridPosition.y - 1)
                       Orientation.WEST -> GridPosition(gridPosition.x - 1, gridPosition.y)
                   }
               }
           }
        }
    }
}

File: solutions/scala/robot-simulator/1/src/main/scala/Robot.scala
class Robot(direction: Bearing.Value, position: Tuple2[Int, Int])  {
    var dir = direction
    var pos = position

    def turnRight: Robot = {
        dir match {
            case Bearing.North => dir = Bearing.East
            case Bearing.East => dir = Bearing.South
            case Bearing.South => dir = Bearing.West
            case Bearing.West => dir = Bearing.North
        }
        this
    }

    def turnLeft: Robot = {
        dir match {
            case Bearing.North => dir = Bearing.West
            case Bearing.East => dir = Bearing.North
            case Bearing.South => dir = Bearing.East
            case Bearing.West => dir = Bearing.South
        }
        this
    }

    def advance: Robot = {
        dir match {
            case Bearing.North => pos = (pos._1, pos._2 + 1)
            case Bearing.East => pos = (pos._1 + 1, pos._2)
            case Bearing.South => pos = (pos._1, pos._2 - 1)
            case Bearing.West => pos = (pos._1 - 1, pos._2)
        }
        this
    }

    def simulate(str: String): Robot = {
        str.map(symbol => {
            symbol match {
                case 'R' => turnRight
                case 'L' => turnLeft
                case 'A' => advance
            }
        })
        this
    }

    def coordinates: (Int, Int) = pos

    def bearing: Bearing.Value = dir

}


object Robot {
    val robot = new Robot(Bearing.North, (0, 0))
    def apply(direction: Bearing.Value, position: Tuple2[Int, Int]): Robot = {
        robot.dir = direction
        robot.pos = position
        robot
    }
}

